### PR TITLE
proxy: Fix false positives in polling-based fs watches

### PR DESCRIPTION
**Pull request by hawkw**
There are currently two issues which can lead to false positives (changes being
reported when files have not actually changed) in the polling-based filesystem
watch implementation. 

The first issue is that when checking each watched file for changes, the loop
iterating over each path currently short-circuits as soon as it detects a 
change. This means that if two or more files have changed, the first time we
poll the fs, we will see the first change, then if we poll again, we will see
the next change, and so on. 

This branch fixes that issue by always hashing all the watched files, even if a
change has already been detected. This way, if all the files change between one
poll and the next, we no longer generate additional change events until a file
actually changes again.

The other issue is that the old implementation would treat any instance of a 
"file not found" error as indicating that the file had been deleted, and 
generate a change event. This leads to changes repeatedly being detected as
long as a file does not exist, rather than a single time when the file's 
existence state actually changes.

This branch fixes that issue as well, by only generating change events on
"file not found" errors if the file existed the last time it was polled. 
Otherwise, if a file did not previously exist, we no longer generate a new 
event.

I've verified both of these fixes through manual testing, as well as a new
test for the second issue. The new test fails on master but passes on this
branch.

Signed-off-by: Eliza Weisman <eliza@buoyant.io>

### CODE REVIEW

File: proxy/src/fs_watch.rs [hunk 1]
```diff
@@ -48,29 +48,22 @@ where
...
-                    Err(ref e) => {
+                let has_changed = file
+                    .has_changed()
+                    .unwrap_or_else(|e| {
                         warn!("error hashing {:?}: {}", &file.path, e);
-                    }
+                        false
+                    });
+                if has_changed {
+                    any_changes = true;
```
Comment (olix0r) [inline proxy/src/fs_watch.rs]: can't we just break the loop at this point?  If so, i'd probably extract this into another fn so you can just return early.

File: proxy/src/fs_watch.rs [hunk 2]
```diff
@@ -122,14 +115,37 @@ impl PathAndHash {
...
-        if changed {
-            self.last_hash.replace(hash);
+        match fs::read(&self.path) {
+            Ok(contents) => {
+                // If we were able to read the file, compare the hash of its
+                // current contents with the previous hash to determine if it
+                // has changed.
+                let hash = Some(digest::digest(&digest::SHA256, &contents[..]));
+                let changed = self.last_hash.borrow().as_ref()
+                    .map(Digest::as_ref) != hash.as_ref().map(Digest::as_ref);
```
Comment (olix0r) [inline proxy/src/fs_watch.rs]: this is pretty hard to parse. can you move the map up onto the prior line?

File: proxy/src/fs_watch.rs [hunk 3]
```diff
@@ -373,6 +394,67 @@ mod tests {
...
+            watch
+        });
+
+        let watch2 = watch.clone();
+        // Check if the stream has become ready every one second.
+        let stream = ::tokio::timer::Interval::new(
+            Instant::now(),
+            Duration::from_secs(1),
+        )
+            .map(|_| {
```
Comment (olix0r) [inline proxy/src/fs_watch.rs]: pretty jarring. perhaps create the interval in its own statement.  Also it's a red flag to see `Instant::now()`.  Shouldn't we be using tokio's clock?

File: proxy/src/fs_watch.rs [hunk 4]
```diff
@@ -48,29 +48,22 @@ where
 
     Interval::new(Instant::now(), interval)
```
Comment (olix0r) [inline proxy/src/fs_watch.rs]: we should probably be using tokio's clock instead of Instant

File: proxy/src/fs_watch.rs (hunk 1 above)
Comment (hawkw) [reply proxy/src/fs_watch.rs]: @olix0r: that's what we were doing previously, but it was causing false positives when more than one file has changed since the last poll. if we return early here, we won't have hashed all the files yet, so we'll see more "changes" on the next poll even if there haven't been any new changes since the last poll.

File: proxy/src/fs_watch.rs (hunk 4 above)
Comment (hawkw) [reply proxy/src/fs_watch.rs]: Yeah, thanks for catching that. I forgot we had upgraded to a tokio version with `Clock`; will change all uses of `Instant::now` in this branch.

File: proxy/src/fs_watch.rs (hunk 2 above)
Comment (hawkw) [reply proxy/src/fs_watch.rs]: finding a way of structuring this that was readable but also placated the borrow checker was kind of tricky; I'll take another crack at it though.

File: proxy/Cargo.toml [hunk 5]
```diff
@@ -36,6 +36,9 @@ regex = "1.0.0"
 # networking
 tokio = "0.1.7"
 tokio-signal = "0.2"
+# We have to depend on `tokio-timer` rather than using the re-exports in the
+# `tokio crate`, as `tokio` doesn't expose the `Clock` trait.
```
Comment (olix0r) [inline proxy/Cargo.toml]: This comment isn't worth its weight, imo.

Preferred:
```
tokio-timer = "0.2.4" # for tokio_timer::clock
```

File: proxy/src/fs_watch.rs [hunk 6]
```diff
@@ -121,15 +114,49 @@ impl PathAndHash {
...
+            Err(ref e) if e.kind() == io::ErrorKind::NotFound => {
+                if self.last_hash.borrow().is_some() {
+                    // If we have a previous hash, then the file was deleted,
+                    // so it has changed.
+                    trace!("{:?} deleted", &self.path);
+                    self.last_hash.replace(None);
+                    Ok(true)
+                } else {
+                    // Otherwise, it didn't exist previously, so there hasn't
+                    // been a change.
```
Comment (olix0r) [inline proxy/src/fs_watch.rs]: This is clear from the context

File: proxy/src/fs_watch.rs [hunk 7]
```diff
@@ -121,15 +114,49 @@ impl PathAndHash {
...
+                // If we were able to read the file, compare the hash of its
+                // current contents with the previous hash to determine if it
+                // has changed.
+                let hash = Some(digest::digest(&digest::SHA256, &contents[..]));
+                let changed = self.last_hash
+                    .borrow()
+                    .as_ref()
+                    .map(Digest::as_ref)
+                    // Compare the last hash with the current hash.
+                    .ne(&hash.as_ref().map(Digest::as_ref));
```
Comment (olix0r) [inline proxy/src/fs_watch.rs]: Can this be written as:

The "compare" comment seems like a bit of a red flag to me -- it's an extremely obvious comment, but it's necessary here because it's a bit hard to read otherwise. This makes me think we should make it more readable so the comment can go away.  Something more like:
```rust
let prior_hash = Digest::as_ref(&digest::digest(&digest::SHA256, &contents[..]);
let current_hash = self.last_hash.borrow().as_ref().map(Digest::as_ref);

let changed = prior_hash.ne(current_hash.as_ref());
if changed {
}
Ok(changed))
```

File: proxy/src/fs_watch.rs [hunk 8]
```diff
@@ -121,15 +114,49 @@ impl PathAndHash {
...
+                if changed {
+                    trace!("{:?} changed", &self.path);
+                    self.last_hash.replace(hash);
```
Comment (olix0r) [inline proxy/src/fs_watch.rs]: extremely nitty: i think `last_hash` is a not-great name -- last can mean _final_ as well as _prior_.  But I believe this is actually best described as the _current_ hash! Because it is.

Comment (hawkw) [reply proxy/src/fs_watch.rs]: Yeah, that's a good point. Will change it to `curr_hash`.

File: proxy/src/fs_watch.rs (hunk 7 above)
Comment (hawkw) [reply proxy/src/fs_watch.rs]: Unfortunately, if I recall correctly, the borrow checker wasn't a fan of doing it like that.
